.circleci/config.yml

```diff
@@ -22,76 +27,201 @@ anchors:
+      branches:
+        only: master
+
+  filter_releases: &filter_releases
+    filters:
+      branches:
+        ignore: /.*/
+      tags:
+        only: /^v.*/
+
+  filter_pr: &filter_pr
+    filters:
+      tags:
+        ignore: /^v.*/
+      branches:
+        ignore: master
+
+# -------------------------
+#   EXECUTORS
+# -------------------------
+executors:
+  # Puppeteer tests run in Chrome
+  browser_docker:
+    <<: *default_environment
+    docker:
+      - image: allofustest/workbench:buildimage-0.0.16
+      - image: circleci/node:13.10.1-browsers
+        environment:
+          NODE_ENV: development
+          CI: true
+
+  # workbench default docker image
+  workbench_docker:
+    <<: *default_environment
     docker:
       - image: allofustest/workbench:buildimage-0.0.16
-      - image: mysql:5.7
+
+  mysql_docker:
+    <<: *default_environment
+    docker:
+      - image: allofustest/workbench:buildimage-0.0.16
+      - image: mysql:5.7 # Is this same as circleci/mysql:5.7 ? This image is not cached on CircleCI
         environment:
           - MYSQL_ROOT_PASSWORD=ubuntu
           - MYSQL_USER=ubuntu
           - MYSQL_PASSWORD=ubuntu
-    working_directory: ~/workbench
     environment:
-      # Drop to ~1.5G; the local API server starts two Java processes (the API
-      # server and the cron emulator). The default machine has 4GB, so we want
-      # the Java memory limit to be below half of that: other things on the
-      # machine need memory as well. Using 2G here manifested in 137 errors.
-      # If this limit proves to be too low, an alternative would be to use a
-      # larger resource class:
-      #   https://circleci.com/docs/2.0/configuration-reference/#resource_class
-      JAVA_TOOL_OPTIONS: -Xmx1536m
-      TERM: dumb
       MYSQL_ROOT_PASSWORD: ubuntu
+
+# -------------------------
+#   COMMANDS
+# https://circleci.com/docs/2.0/reusing-config/#authoring-reusable-commands
+# -------------------------
+commands:
+  activate_service_accont_creds:
+    description: "Activiate CircleCI service account credential"
+    steps:
+      - run:
+          name: Activate CircleCI Service Account credential
+          working_directory: ~/workbench
+          # Used to call gsutil from the circle environment.
+          command: ci/activate_creds.sh api/circle-sa-key.json
+
+  git_checkout:
+    description: "git checkout and update submodules"
     steps:
       - checkout
-      - run: *update_git_submodules
+      - run:
+          name: Update Git Submodules
+          command: git submodule update --init --recursive
+
+  ensure_branch_has_api_changes:
+    description: "Ensure branch has changes to API Code"
+    steps:
+      - run:
+          name: Ensure Branch has Changes to API Code
+          command: |
+            if [ ${CIRCLE_BRANCH} != "" ] && [ ${CIRCLE_BRANCH} != "master" ] && [ $(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep api/ | wc -l | xargs) == 0 ]; then
```

Comment:
nit: if this gets any longer, can we move it to an include file or actually deploy it as a script? Shell code inside yaml is a bit difficult to read.

Reply:
original code. I don't want to change and break it.

Reply:
Right, just wondering if you came across a way to do that.